### DISCUSSION TITLE
Class based websocket - on_receive and on_disconnect not working

### DISCUSSION THREAD

**Discussion by asaff1** [fastapi/fastapi#8898] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example
I followed the example here: 
https://www.starlette.io/endpoints/#websocketendpoint

```Python
router = APIRouter()

class WebSocketHandler(WebSocketEndpoint):
    encoding = "text"

    async def on_connect(self, websocket):
        await websocket.accept()
        while True:
            await websocket.send_json({"hello": "world"})
            await asyncio.sleep(5)

    async def on_receive(self, websocket, data):
        print(f"websocket received: {data}")

    async def on_disconnect(self, websocket, close_code):
        print(f"websocket client disconnected, code={close_code}")
```

Tried to register with:
```Python
router.add_websocket_route("/ws", WebSocketHandler)
```
Or to put the decorator `@router.websocket_route("/ws")` on the class. Not working either.

In both cases, only `on_connect` gets called. The other two won't.
The test client is simple JS code:
```
ws = new WebSocket(`ws://${location.host}/api/v1/ws`);
ws.onmessage = (ev) => console.log(ev)
```

And then:
```
ws.send("hello")
ws.close()
```

### Description
Client does receive messages, but server does not fire `on_receive` and `on_disconnect`

### Environment

* OS: Windows 
* FastAPI Version [e.g. 0.3.0]: 0.63.0

* Python version: 3.8.9

**Comment by naufalafif** [comment]
as alternative, u can also do someting like this

```python
from fastapi import APIRouter, WebSocketDisconnect, WebSocket

router = APIRouter()

@router.websocket("/")
async def lab_socket(websocket: WebSocket):

    async def on_connect():
        # websocket connect event
        await websocket.accept()

    async def on_receive(data):
        # websocket receive event
        await websocket.send_text(f"Message text was: {data}")

    async def on_disconnect():
        # websocket diconnect event
        pass

    try:
        await on_connect()
        while True:
            data = await websocket.receive_text()
            await on_receive(data)
    except WebSocketDisconnect:
        await on_disconnect()
```

**Comment by parkerbjur** [comment]
Is there any more guidance on getting classed based websockets to work?

**Comment by HolyShaq** [comment, marked as answer]
I use this implementation to make class-based websockets work.

Base websocket class:
```python
class WebsocketBase:
    def __init__(self, websocket: WebSocket):
        self.websocket = websocket

    async def on_connect(self):
        pass

    async def on_receive(self, data):
        pass

    async def on_disconnect(self):
        pass

    async def handle(self):
        # Handle connect event
        await self.on_connect()
        try:

            # Handle on receive events
            while True:
                data = await self.websocket.receive_text()
                await self.on_receive(data)

        # Handle disconnect event
        except WebSocketDisconnect:
            await self.on_disconnect()
```

Custom decorator you will use in place of `@router.websocket("/")`
```python
def websocket_wrapper(router: APIRouter, path: str):
    def decorator(cls: Type[WebsocketBase]):

        @router.websocket(path)
        async def wrapper(websocket: WebSocket):
            instance = cls(websocket)
            await instance.handle()

        return wrapper

    return decorator
```

Then, you can define websockets like this:
```python
@websocket_wrapper(router, "/global/")
class GlobalMessagesWebsocket(WebsocketBase):
    async def on_connect(self):
        print("Connected")
        await self.websocket.accept()

    async def on_receive(self, data):
        print(f"Received: {data}")
        await self.websocket.send_text(f"Message text was {data}")

    async def on_disconnect(self):
        print("Disconnected")
        pass
```
